### PR TITLE
#1259 Add GroupPushRules and GroupPushRulesManager classes

### LINKED ISSUE #1259: Manage push rules on a group
## Description of the problem, including code/CLI snippet

It appears that a groups object doesn't have any methods to read or modify push rules

https://docs.gitlab.com/ee/api/groups.html#push-rules

## Expected Behavior

Expected similar functionality to the [projects.pushrules](https://python-gitlab.readthedocs.io/en/stable/gl_objects/projects.html#id21) methods, to be able to read and modify push rules at the group level that projects underneath the group will inherit.

## Actual Behavior

No pushrules methods exist in a groups object

## Specifications

  - python-gitlab version: python-gitlab==2.5.0
  - API version you are using (v3/v4): v4
  - Gitlab server version (or gitlab.com): gitlab.com

### PR DESCRIPTION
**Pull request by gokeefe**
Add classes to read and update push rules on groups

Closes #1259.

### CODE REVIEW

File: gitlab/v4/objects/__init__.py [hunk 1]
```diff
@@ -1350,6 +1350,44 @@ class GroupProjectManager(ListMixin, RESTManager):
...
+        (
+            "deny_delete_tag",
+            "member_check",
+            "prevent_secrets",
+            "commit_message_regex",
+            "branch_name_regex",
+            "author_email_regex",
+            "file_name_regex",
+            "max_file_size",
+        ),
```
Comment (nejch) [inline gitlab/v4/objects/__init__.py:1387]: Could you please check the existing list of arguments available and update this list? https://docs.gitlab.com/ce/api/groups.html#add-group-push-rule just a few missing.
Thanks :) (even though one if them is missing from the response per your comment, and I realize this means our project push rules are probably outdated ... I've opened https://github.com/python-gitlab/python-gitlab/issues/1282)

**Review by nejch — CHANGES_REQUESTED** [review]
Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?

We can't really run integration tests at the moment on EE features but if you want to give it a shot you can add some unit tests for this in `gitlab/tests/objects`. You can look at some existing examples at https://github.com/python-gitlab/python-gitlab/tree/master/gitlab/tests/objects, and use the example responses at https://docs.gitlab.com/ce/api/groups.html#push-rules for mocks.

I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.

**Comment by gokeefe** [comment]
> Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?

Yes, right here: https://gitlab.com/gitlab-org/gitlab/-/issues/297559

> I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.

Awesome, the travis-ci build failed and I had no idea why because the same tests were working fine locally. I'll have a look at the rest next week. Thanks!

**Comment by nejch** [comment]
> > Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?
> 
> Yes, right here: https://gitlab.com/gitlab-org/gitlab/-/issues/297559
> 
> > I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.
> 
> Awesome, the travis-ci build failed and I had no idea why because the same tests were working fine locally. I'll have a look at the rest next week. Thanks!

Just FYI - we've split the objects package into modules now so this will have to be rebased and moved into `gitlab/v4/objects/push_rules.py` and the manager also imported in `groups.py`. Take a look at https://github.com/python-gitlab/python-gitlab/pull/1304 for an example on how to add it.

I'll assign this to you for now but if you need help just let me know :)

File: gitlab/v4/objects/__init__.py (hunk 1 above)
Comment (gokeefe) [reply gitlab/v4/objects/__init__.py]: That's done. I also updated the ProjectPushRulesManager with the missing attributes. Not sure if you'd prefer that in a separate pull request.

Comment (gokeefe) [reply gitlab/v4/objects/__init__.py]: I'm not too familiar with pytest so i'm going to have to work on the unit tests a bit more

**Comment by nejch** [comment]
Hi @gokeefe are you still interested in finishing this?

I think we're ok without pytest and I can add them later if you're not comfortable doing that. This just needs a rebase to resolve some conflicts. Let me know!

**Comment by JohnVillalovos** [comment]
@nejch I updated this to squash it down to one commit and clean things up a little bit. If you have time to review I can work on it some more if needed.

File: docs/gl_objects/groups.rst [hunk 2]
```diff
@@ -376,3 +376,37 @@ Delete a group hook::
...
+* GitLab API: https://docs.gitlab.com/ee/api/groups.html#push-rules
+
+Examples
+---------
+
+Create group push rules (at least one rule is necessary)::
+
+    group.pushrules.create({'deny_delete_tag': True})
+
+Get group push rules (returns None is there are no push rules)::
```
Comment (nejch) [inline docs/gl_objects/groups.rst:401]: (Edit: ah I realize this is also wrong in the project docs, nvm)

```suggestion
Get group push rules (returns ``None`` if there are no push rules)::
```

**Review by nejch — APPROVED** [review]
Thanks a lot @gokeefe for the PR and @JohnVillalovos for bringing this up to date! :)

**Pull request merged**